# Evolver Loop 2 Analysis: Understanding B3 Failure & Next Steps

## Goal
Analyze why EfficientNet-B3 with enhanced augmentation performed worse than B0 baseline, and identify high-ROI improvements.

## Key Questions
1. What caused the performance degradation in exp_003?
2. How can we build on the successful B0 baseline?
3. What ensemble strategies will give best ROI?

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import json

# Load session state to review experiments
with open('/home/code/session_state.json', 'r') as f:
    session_state = json.load(f)

experiments = session_state['experiments']
print("Experiment Summary:")
print("=" * 50)
for exp in experiments:
    print(f"{exp['name']}: {exp['score']:.6f} (model: {exp['model_type']})")
    print(f"  Notes: {exp['notes'][:200]}...")
    print()

print(f"\nGold threshold: 0.038820")
print(f"Best score: {min([exp['score'] for exp in experiments]):.6f}")
print(f"Improvement needed: {0.038820 - min([exp['score'] for exp in experiments]):.6f}")

Experiment Summary:
002_baseline_efficientnet: 0.020219 (model: efficientnet_b0)
  Notes: Baseline EfficientNet-B0 model with transfer learning from ImageNet. Used PyTorch implementation with proper GPU acceleration on NVIDIA A100. Training: 15 epochs, batch size 64, Adam optimizer with le...

exp_003_efficientnet_b3_enhanced: 0.051978 (model: efficientnet-b3)
  Notes: Enhanced EfficientNet-B3 with early stopping (patience=5), advanced augmentation (CutMix, MixUp), and improved regularization (dropout=0.3, label smoothing=0.1). Result: 0.051978 validation log loss -...


Gold threshold: 0.038820
Best score: 0.020219
Improvement needed: 0.018601


## Analysis: Why Did B3 Fail?

The B3 experiment degraded from 0.020219 to 0.051978 - a 157% worse performance. Let's analyze potential causes:

1. **CutMix/MixUp too aggressive**: These techniques mix images and labels, which can hurt performance on clean, balanced datasets
2. **Learning rate suboptimal**: Same LR=1e-4 may not work for larger model
3. **Overfitting despite regularization**: B3 has 11M params vs B0's 5M - more capacity to overfit
4. **Early stopping didn't trigger**: Trained all 20 epochs, may have overfitted

Key insight: The baseline B0 already beats gold significantly. We should:
- Build on what works (B0 architecture)
- Add proven improvements (TTA, early stopping, simple ensembles)
- Avoid overcomplicating with aggressive augmentation